kendzi3d-plugin/src/main/java/kendzi/josm/kendzi3d/jogl/model/BarrierFence.java

```diff
@@ -127,6 +132,9 @@ public void buildWorldObject() {
         model = modelBuilder.toModel();
         model.setUseLight(true);
         model.setUseTexture(true);
+        model.setUseTextureAlpha(true);
+        model.setUseCullFaces(!PREFER_TWO_SIDED.get());
+        model.setUseTwoSidedLighting(PREFER_TWO_SIDED.get());
 
```

Comment:
You can actually check that this works without changing the code (just by toggling the preference in josm config):

- with face culling the texture on the other side is mirrored

The fence texture is not completely symmetrical, the spacing between the post and the picket is for a tiny amount larger on the left side than on the right side.  If you zoom close enough, you will notice.  Moving the viewport to the other side of the fence, larger and smaller gap switch sides from left to right.

- with two sided lighting this does not happen, the gaps remain at their position, no matter from which side the fence is viewed  (which is more natural with respect to reality, but barely noticable;  so less so that it is not an argument to not employ face culling as a standard method in case two sided lighting does not work on a specific machine)

Reply:
Don't known why it happen but after first time check "Force two site lighting" fence became lighter, and stay like that even if I un-select that option.

Reply:
preference in josm config  !=  "force two sided lighting"

There is
* menu option  "force two sided lighting"  in 3dMenu

* PREFER_TWO_SIDED = new BooleanProperty("kendzi3d.models.preferTwoSidedLightingOverFaceCulling", true);  in AbstractModel.java


The menu option ''forces'' two sided lighting for ''all'' models  in ''any'' case ("dumb" debug option).

PREFER_TWO_SIDED  can be toggled ''only'' is josm preferences (F12 -> expert options -> kendzi3d.models.preferTwoSidedLightingOverFaceCulling)  and is more intelligent than the menu option above.   If it is enabled, it will turn on two sided lighting for ''some'' select models only, that have alternative methods to draw themselves in the source code.

Chronologically, PREFER_TWO_SIDED option was added after menu option.  PREFER_TWO_SIDED option is evaluated by BarrierFence.java and BarrierWall.java.   This means if you want to toggle drawing methods for fences and walls  (1st method: two sided lighting;  2nd method:  two-faced, with face culling),  you _need_ to
* turn off the forcing menu option
* toggle PREFER_TWO_SIDED   in  josm preferences


I have made  the forcing menu option  show only in expert mode of josm, but maybe it is better to not view it at all.  It really is a ''legacy'' option  that was introduced  before the prefer_two_sided property.

Reply:
Note that toggling PREFER_TWO_SIDED  will _not_ give you noticeable difference in lighting.  It should give you exactly the same lighting result (if the drivers are OK and support both methods).  However there _is_ a perceptible difference when looking closely at how the texture images are applied on the different sides of the fence.

Looking for this difference in detail, you can visually examine which drawing method is in effect.  I'll try to add a picture of this, if I find time.

Reply:
I think I understand how PREFER_TWO_SIDED is working. I simply make stupid click test on menu option "force two sided lighting". So I started with that option un-selected (default setting):
![bug1_unselected](https://user-images.githubusercontent.com/811830/34470222-50348f72-ef2d-11e7-8c7d-75f0bc0b9984.png)
Now I select that option, nothing should changed on fence as it is already rendered as lighted from both site (PREFER_TWO_SIDED is on default value), but I see that fence was changed. It become lighter:
![bug1_selected](https://user-images.githubusercontent.com/811830/34470237-9ccf0510-ef2d-11e7-9bcd-52f4202398f4.png)
After un-selecting option  "force two sided lighting" in menu, nothing is changing fence is still lighter then it was initially. To reset it to initial color i need to close/open 3d view.

Reply:
Ok, so you expect *if* PREFER_TWO_SIDED is on default value, *then* toggling the force two sided menu option should not have any effect  on fence drawing, correct?

It seems as if the ground and wall in the second image are also brighter when compared to the one above, so brighter fence *may* be a result from additional indirect lighting.  Let this be the case, then one argument remains:  Turning the menu option *off*, rendering should yield the dim (or dimmer) lighting again.

Currently this is the case only after window is reopened, but it should toggle without being closed.  Is this the specific issue you are referring to?

Or are you rather curious why the model is being drawn brighter when the menu option is *on* (enabled), when there seemingly should be no difference (you could argue indirect lighting may not lead to a difference in lighting this big)?

Reply:
Looks like only fence is affected. Download that images and open them in image viewer one after another, because on comment second appear lighter but it is not.

So as I understand indirect lighting should not be changed at all, fence model have TwoSidedLighting flag set to true, and on ModelRender there is "or" condition on flag in model OR global flag. So if model have that flag set the global flag will be ignored (it is "or" condition)
``` (model.useTwoSided || drawTwoSided) ```
By looking into code I'm surprised that fence color is changed in anyway. More if it change ( assuming it should change) I'm surprised that it is not returning to old value after that global setting is un-selected. Not returning to old value after selecting/un-selecting looks for me as bug.
So I'm curious why the model is being drawn brighter when the menu option is on (enabled), when in my opinion should be no difference. If that is ok behavior please explain why.

Turning 3d window off/on "fix" lightning because opengl is reinitialized with default values, so looks like something is not clear somewhere after un-selecting "force two sided lighting" in global.

Reply:
Actually I cannot reproduce this here.  You can see that the street back side receives a different lighting in the second image, so the force option toggled correctly to enabled state, but the fence drawing is completely unaffected (as you would expect).

Data was chosen so both sides of the fence drawing can be seen:

![prefer-two-sided-enabled_force-two-sided-disabled_default-after-initial-view](https://user-images.githubusercontent.com/10984760/34471420-751fd7c6-ef49-11e7-9481-bdde2c436c4c.png)
![prefer-two-sided-enabled_force-two-sided-enabled](https://user-images.githubusercontent.com/10984760/34471423-77f3c156-ef49-11e7-9261-972e863c39b5.png)

Can you post your light source parameters and a data link for me to try with the same data you've used?

Reply:
To reproduce you need to watch "back" side of fence which is lighted. I'm attaching test data, look from south to north. Start with global "force two sided lighting" un-selected.

Looks like opengl keeps track of front and back component separably. At least as I understand it. You sometime set both of them by using flag GL_FRONT_AND_BACK and sometime you are set only front material by flag GL_FRONT. But after setting front and back, the back color is never reset to its default values.

[light_bug1.osm.txt](https://github.com/kendzi/kendzi3d/files/1596445/light_bug1.osm.txt)

Reply:
Ok, I can reproduce with light_bug1.osm

But I can't follow your explanation, because GL_FRONT_AND_BACK is used at all times to set fence material as long as PREFER_TWO_SIDED is enabled, see this line in ModelRender.java
```java
setupMaterial2(gl, material, (model.useTwoSided || drawTwoSided) ?  GL.GL_FRONT_AND_BACK : GL.GL_FRONT);
```

* model.useTwoSided is set for fence model if PREFER_TWO_SIDED is enabled
* drawTwoSided  corresponds to force two sided lighting menu option

I will have a further look at it, but maybe you find it first..